### ISSUE TITLE
Test accessible drop down menu

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/a11y-research#13]
For accessibility, the PhET Menu should behave like a Dropdown menu.  This is different than a Dialog in that there is no close button, it does not restrict navigation, and the menu will close once the user leaves the menu.

Here is an example:
https://staff.washington.edu/tft/tests/menus/simplyaccessible/index.html

In that example, the drop down menu is represented as:
```html
<div role="navigation" aria-label="Main menu">
	<ul class="nav" role="menubar" aria-hidden="false">
		<li role="menuitem" aria-haspopup="true">
			<a href="">About</a>
			<ul data-test="true" aria-hidden="false" role="menu" class="show-menu">
				<li role="menuitem"><a href="" tabindex="0">News</a></li>
				<li role="menuitem"><a href="" tabindex="0">Governance</a></li>
				<li role="menuitem"><a href="" tabindex="0">Diversity</a></li>
				<li role="menuitem"><a href="" tabindex="0">ContactUs</a></li>
			</ul>
		</li>
		<li role="menuitem" aria-haspopup="true">
			<a href="">Academics</a>
			<ul data-test="true" aria-hidden="true" role="menu">
				<li role="menuitem"><a href="" tabindex="-1">Degree Programs</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Faculty</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Distance Learning</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Libraries</a></li>
			</ul>
		</li>
		<li role="menuitem" aria-haspopup="true">
			<a href="">Admissions</a>
			<ul data-test="true" aria-hidden="true" role="menu">
				<li role="menuitem"><a href="" tabindex="-1">Undergraduate</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Tuition</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Financial Aid</a></li>
			</ul>
		</li>
		<li role="menuitem" aria-haspopup="true">
			<a href="">Visitors</a>
			<ul data-test="true" aria-hidden="true" role="menu" class="">
				<li role="menuitem"><a href="" tabindex="-1">Events</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Campus Map</a></li>
				<li role="menuitem"><a href="" tabindex="-1">Parking</a></li>
			</ul>
		</li>
	</ul>	
</div>
```

It should be tested with screen readers.  There are a few comments on the example page that suggest there could be a few issues like

> The arrow keys work great without a screen reader, but if JAWS is running it seems to intercept keystrokes and not release them to JavaScript.

> The original menu (Option 6) doesn't work as well as I would expect it to with screen readers.

**Comment by jessegreenberg** [comment]
I am not sure that nested menu items are working here, In NVDA and JAWS I hear something like

> menu item,  subMenu, menu, menu item, link, News

when the focus is on the "News" link in the first drop down menu.  This seems like too much repetition.

**Comment by jessegreenberg** [comment]
Chrome doesn't support ```role=menu``` with NVDA but it works very well in firefox.  Example PhET Menu in a JSFiddle:
https://jsfiddle.net/2un2cv0g/2/show/

https://www.w3.org/WAI/GL/wiki/Using_ARIA_menus#ARIA_Menu_Widget_Technique shows an example and a "success criterion" for menus.

**Comment by jessegreenberg** [comment]
Hey there is a MenuButton!

https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton

The keyboard interaction:
>With focus on the button:
Space and Enter: open the menu and place focus on the first menu item.
(Optional) Down Arrow: opens the menu and moves focus to the first menu item.
The keyboard behaviors needed after the menu is open are described in 2.19 Menu or Menu bar.

**Comment by jessegreenberg** [comment]
It is part of the ARIA 1.1 spec though, so support might be pretty poor.

**Comment by jessegreenberg** [comment]
It is just an element with role=button and aria-haspopup set to true.  Screen readers should then interperet this as a Menu Button.

Here is a test:
https://jsfiddle.net/2un2cv0g/3/show/

It works great with Firefox + NVDA

**Comment by jessegreenberg** [comment]
JAWS doesnt work very well with the above example, maybe because it is because it is struggling in the JSFiddle frame.

**Comment by jessegreenberg** [comment]
JAWS recognizes the aria-haspopup attribute, and calls the button a "Menu Button".  But the arrow keys do not work.  That is probably not critical, as long as the virtual cursor works and the user can tab as they are used to, it shouldn't be a problem.

**Comment by jessegreenberg** [comment]
@terracoda what do you think about something like this for the PhET Menu:

```html
  <button id="phet-button" aria-haspopup>PhET Menu</button>
  <ul id = "phet-menu" hidden role="menu">
    <li role="menuitem"><button href="">Options...</button></li>
    <li role="menuitem"><a href="../..">PhET Website...</a></li>
    <li role="menuitem"><a href="">Report a Problem...</a></li>
    <li role="menuitem"><a href="">Check for Updates</a></li>
    <li role="menuitem"><button href="">Screenshot</button></li>
    <li role="menuitem"><button href="">Full Screen</button></li>
    <li role="menuitem"><button href="">About...</button></li>
  </ul>
```

**Comment by terracoda** [comment]
@jessegreenberg, that looks good, and maybe that first button is a menuitem of the main menubar? And for the _Full Screen_ maybe we shoudl try the role menuitemcheckbox?

**Comment by terracoda** [comment]
@jessegreenberg, interesting that the dot dot dot is also mentioned as a convention for pop-ups. I didn't realize that :-)

**Comment by terracoda** [comment]
@jessegreenberg I don't hear any difference between `menuitem` and `menuitemcheckbox` with VoiceOver and Safari.

**Comment by terracoda** [comment]
@jessegreenberg, I combined my code with yours and it seems to sound pretty good in VO in FF and Safari.

**Comment by jessegreenberg** [comment]
> maybe that first button is a menuitem of the main menubar

Sounds great!

> for the Full Screen maybe we shoudl try the role menuitemcheckbox

Sure, or maybe we could use a visual checkbox in the menu? in that case, it could be ```<input type="checkbox" role="menuitem">``` Either way, this might still add information if ```menuitemcheckbox``` is not supported.

> the dot dot dot is also mentioned as a convention for pop-ups

That is interesting, where is that mentioned?  I found it annoying that the '...' was read by AT, but maybe it is informative for regular AT users.

**Comment by terracoda** [comment]
@jessegreenberg I read the "dot dot dot" thing this morning, but I haven't re-located it yet. I think it was in the ARIA spec  (1.0 maybe) regarding menu items and pop-ups.

**Comment by terracoda** [comment]
@jessegreenberg, found it :-) in the [ARIA 1.1 Authoring Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

> **2.6 Button**
A button is a widget that enables users to trigger an action or event, such as submitting a form, opening a dialog, canceling an action, or performing a delete operation. A common convention for informing users that a button launches a dialog is to append "…" (ellipsis) to the button label, e.g., "Save as…".

**Comment by jessegreenberg** [comment]
Excellent thanks @terracoda! That is really helpful.

**Comment by terracoda** [comment]
@jessegreenberg, just gathering links to potentially related documents into this issue. 

You were asking about the parent "menu bar" or "tool bar" in yesterday's meeting, right? This "thing" will contain the Keyboard Help Button, the PhET Menu Button and eventually the Tabs for multi-screen sims.

The closed Issue #11 links to a [google doc](https://docs.google.com/document/d/12jd7xKEt8P_EUwTrbhLuvIKkZEDaqdFgyQ-EXeLgGkE/edit#heading=h.j931sqqmwrdk) and some [sample code](https://github.com/terracoda/phet-templates/blob/master/a11y-menu-template.html) that I started with.

I closed issue #11, when I realized you were working on the same thing in issue #13.

Issue #12 is related as well, but was specifically about the PhET menu changes.

**Comment by terracoda** [comment]
I updated a sample html template for the sim’s bottom bar.
I used a footer element instead of a tool or menu bar. This might be more reliable and provide a nice landmark.

I've used some nav elements with aria-labels, so it sounds pretty good in Voice Over, but it’s still quite rough. The PhET Menu buttons are not being read out yet (when expanded), and are not navigable with arrow keys (when expanded), so I think I have done something wrong there. I've adapted some tips from Pickering's _Inclusive Design Patterns_.

Also, I am not 100% sure if sim screen navigation should go inside or outside this footer. This decision does not have to happen immediately, but soon. 

The sample template is in my own repo at this link:
https://github.com/terracoda/phet-templates/blob/master/a11y-menu-template.html

@jesse, also, is this the active issue for this topic, or is issue #12, the proper place?

**Note:** Pickering has an nice pattern for using SVG on the button icon. I haven't included that as we likely do not need it in the hidden PDOM. It might be useful for other situations, though.

**Comment by terracoda** [comment]
I think this is a stale issue and structure was addressed in issues #25 and #215, so I am closing this issue.
